qiskit/algorithms/gradients/base_sampler_gradient.py

```diff
@@ -91,17 +114,121 @@ def _run(
+            circuits, parameter_values, parameter_sets
+        ):
+            circuit_key = _circuit_key(circuit)
+            if circuit_key not in self._gradient_circuit_cache:
+                transpiled_circuit = transpile(
+                    circuit, basis_gates=supported_gates, optimization_level=0
+                )
+                self._gradient_circuit_cache[circuit_key] = _assign_unique_parameters(
+                    transpiled_circuit
+                )
+            gradient_circuit = self._gradient_circuit_cache[circuit_key]
+            g_circuits.append(gradient_circuit.gradient_circuit)
+            g_parameter_values.append(
+                _make_gradient_parameter_values(circuit, gradient_circuit, parameter_value_)
+            )
+            g_parameter_sets.append(_make_gradient_parameter_set(gradient_circuit, parameter_set))
+        return g_circuits, g_parameter_values, g_parameter_sets
+
+    def _postprocess(
+        self,
+        results: SamplerGradientResult,
+        circuits: Sequence[QuantumCircuit],
+        parameter_values: Sequence[Sequence[float]],
+        parameter_sets: Sequence[set[Parameter] | None],
+    ) -> SamplerGradientResult:
+        """Postprocess the gradient. This computes the gradient of the original circuit from the
+        gradient of the gradient circuit by using the chain rule.
+
+        Args:
+            results: The results of the gradient of the gradient circuits.
+            circuits: The list of quantum circuits to compute the gradients.
+            parameter_values: The list of parameter values to be bound to the circuit.
+            parameters: The sequence of parameters to calculate only the gradients of the specified
+                parameters.
+
+        Returns:
+            The results of the gradient of the original circuits.
+        """
+        gradients, metadata = [], []
+        for idx, (circuit, parameter_values_, parameter_set) in enumerate(
+            zip(circuits, parameter_values, parameter_sets)
+        ):
+            gradient_circuit = self._gradient_circuit_cache[_circuit_key(circuit)]
+            g_parameter_set = _make_gradient_parameter_set(gradient_circuit, parameter_set)
+            # Make a map from the gradient parameter to the respective index in the gradient.
+            parameter_indices = [param for param in circuit.parameters if param in parameter_set]
+            g_parameter_indices = [
+                param
+                for param in gradient_circuit.gradient_circuit.parameters
+                if param in g_parameter_set
+            ]
+            g_parameter_indices = {param: i for i, param in enumerate(g_parameter_indices)}
+            # Compute the original gradient from the gradient of the gradient circuit
+            # by using the chain rule.
+            gradient = []
+            for parameter in parameter_indices:
+                grad_dist = np.zeros(2**circuit.num_qubits)
```

Comment:
It would be good to avoid creating this huge vector by default 🙂

Reply:
I agree. I updated the code to use dict instead of a large vector.